In [84]:
import numpy as np
import pandas as pd

In [85]:
%%HTML
<style type="text/css">
table.dataframe td, table.dataframe th {
border: 1px black solid !important;
color: black !important;
}

#### Load the data from the file

In [86]:
df = pd.read_csv('titanic.csv')
df.head()

Unnamed: 0 Cabin Embarked     Fare  Pclass            Ticket   Age  \
0           0   NaN        S   7.2500       3         A/5 21171  22.0   
1           1   C85        C  71.2833       1          PC 17599  38.0   
2           2   NaN        S   7.9250       3  STON/O2. 3101282  26.0   
3           3  C123        S  53.1000       1            113803  35.0   
4           4   NaN        S   8.0500       3            373450  35.0   

                                                Name  Parch     Sex  SibSp  \
0                            Braund, Mr. Owen Harris      0    male      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      0  female      1   
2                             Heikkinen, Miss. Laina      0  female      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)      0  female      1   
4                           Allen, Mr. William Henry      0    male      0   

   Survived  
0       0.0  
1       1.0  
2       1.0  
3       1.0  
4       0.0

#### Indexing & Selecting data

In [87]:
# Selecting individual column by using column label
df['Age']

0       22.0
1       38.0
2       26.0
3       35.0
4       35.0
        ... 
1304     NaN
1305    39.0
1306    38.5
1307     NaN
1308     NaN
Name: Age, Length: 1309, dtype: float64

In [88]:
df.Age

0       22.0
1       38.0
2       26.0
3       35.0
4       35.0
        ... 
1304     NaN
1305    39.0
1306    38.5
1307     NaN
1308     NaN
Name: Age, Length: 1309, dtype: float64

#### Select multiple columns at once

In [89]:
df[['Name', 'Parch', 'Sex']]

Name  Parch     Sex
0                               Braund, Mr. Owen Harris      0    male
1     Cumings, Mrs. John Bradley (Florence Briggs Th...      0  female
2                                Heikkinen, Miss. Laina      0  female
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)      0  female
4                              Allen, Mr. William Henry      0    male
...                                                 ...    ...     ...
1304                                 Spector, Mr. Woolf      0    male
1305                       Oliva y Ocana, Dona. Fermina      0  female
1306                       Saether, Mr. Simon Sivertsen      0    male
1307                                Ware, Mr. Frederick      0    male
1308                           Peter, Master. Michael J      1    male

[1309 rows x 3 columns]

#### Selecting one or more rows using iloc (Index Locator)

In [90]:
df.iloc[0]

Unnamed: 0                          0
Cabin                             NaN
Embarked                            S
Fare                             7.25
Pclass                              3
Ticket                      A/5 21171
Age                                22
Name          Braund, Mr. Owen Harris
Parch                               0
Sex                              male
SibSp                               1
Survived                            0
Name: 0, dtype: object

In [91]:
df.iloc[[0,1,2]]

Unnamed: 0 Cabin Embarked     Fare  Pclass            Ticket   Age  \
0           0   NaN        S   7.2500       3         A/5 21171  22.0   
1           1   C85        C  71.2833       1          PC 17599  38.0   
2           2   NaN        S   7.9250       3  STON/O2. 3101282  26.0   

                                                Name  Parch     Sex  SibSp  \
0                            Braund, Mr. Owen Harris      0    male      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      0  female      1   
2                             Heikkinen, Miss. Laina      0  female      0   

   Survived  
0       0.0  
1       1.0  
2       1.0

#### Get a list of columns

In [92]:
df.columns

Index(['Unnamed: 0', 'Cabin', 'Embarked', 'Fare', 'Pclass', 'Ticket', 'Age',
       'Name', 'Parch', 'Sex', 'SibSp', 'Survived'],
      dtype='object')

#### Slicing columns

In [93]:
df[df.columns[2:6]]

Embarked      Fare  Pclass              Ticket
0           S    7.2500       3           A/5 21171
1           C   71.2833       1            PC 17599
2           S    7.9250       3    STON/O2. 3101282
3           S   53.1000       1              113803
4           S    8.0500       3              373450
...       ...       ...     ...                 ...
1304        S    8.0500       3           A.5. 3236
1305        C  108.9000       1            PC 17758
1306        S    7.2500       3  SOTON/O.Q. 3101262
1307        S    8.0500       3              359309
1308        C   22.3583       3                2668

[1309 rows x 4 columns]

#### Get number of rows in a dataset

In [94]:
len(df)

1309

#### Get the value from 5th row and Name column in various ways

In [95]:
df.iloc[5]['Name'] # Row centric method

'Moran, Mr. James'

In [96]:
df.iloc[5].Name #  Row centric method

'Moran, Mr. James'

In [97]:
df['Name'][5] # Column centric method

'Moran, Mr. James'

In [98]:
df.Name[5] # Column centric method

'Moran, Mr. James'

##### IN THIS LESSON WE LEARNED 
<LI> LOAD DATA <LI> SELECT SPECIFIC ITEMS FROM DATAFRAME

#### Advanced Indexing and Selection

##### <ul> 
Task 1: create list of passengers names & ages for those passengers under age of 21 

In [99]:
df[df.Age < 21][['Name','Age']]

Name   Age
7                     Palsson, Master. Gosta Leonard   2.0
9                Nasser, Mrs. Nicholas (Adele Achem)  14.0
10                   Sandstrom, Miss. Marguerite Rut   4.0
12                    Saundercock, Mr. William Henry  20.0
14              Vestrom, Miss. Hulda Amanda Adolfina  14.0
...                                              ...   ...
1283                   Abbott, Master. Eugene Joseph  13.0
1286  Smith, Mrs. Lucien Philip (Mary Eloise Hughes)  18.0
1294                          Carrau, Mr. Jose Pedro  17.0
1296    Nourney, Mr. Alfred (Baron von Drachstedt")"  20.0
1300                       Peacock, Miss. Treasteall   3.0

[249 rows x 2 columns]

#### <ul> 
Task 2 How many child passengers ?

In [100]:
print(len(df[df.Age < 21]))

249


#### <ul> 
Task 2 How many passengers are between age 21 and 30

In [101]:
len(df.loc[(df.Age > 21) & (df.Age < 30)])

279

#### <ul> 
Task 3 Find first or third class ticket holders

In [102]:
df[(df.Pclass == 1) | (df.Pclass == 3) ]

Unnamed: 0 Cabin Embarked      Fare  Pclass              Ticket   Age  \
0              0   NaN        S    7.2500       3           A/5 21171  22.0   
1              1   C85        C   71.2833       1            PC 17599  38.0   
2              2   NaN        S    7.9250       3    STON/O2. 3101282  26.0   
3              3  C123        S   53.1000       1              113803  35.0   
4              4   NaN        S    8.0500       3              373450  35.0   
...          ...   ...      ...       ...     ...                 ...   ...   
1304        1304   NaN        S    8.0500       3           A.5. 3236   NaN   
1305        1305  C105        C  108.9000       1            PC 17758  39.0   
1306        1306   NaN        S    7.2500       3  SOTON/O.Q. 3101262  38.5   
1307        1307   NaN        S    8.0500       3              359309   NaN   
1308        1308   NaN        C   22.3583       3                2668   NaN   

                                                   Name  Parch     Sex  SibSp  \
0                               Braund, Mr. Owen Harris      0    male      1   
1     Cumings, Mrs. John Bradley (Florence Briggs Th...      0  female      1   
2                                Heikkinen, Miss. Laina      0  female      0   
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)      0  female      1   
4                              Allen, Mr. William Henry      0    male      0   
...                                                 ...    ...     ...    ...   
1304                                 Spector, Mr. Woolf      0    male      0   
1305                       Oliva y Ocana, Dona. Fermina      0  female      0   
1306                       Saether, Mr. Simon Sivertsen      0    male      0   
1307                                Ware, Mr. Frederick      0    male      0   
1308                           Peter, Master. Michael J      1    male      1   

      Survived  
0          0.0  
1          1.0  
2          1.0  
3          1.0  
4          0.0  
...        ...  
1304       NaN  
1305       NaN  
1306       NaN  
1307       NaN  
1308       NaN  

[1032 rows x 12 columns]

#### <ul> 
Task 4 Find passengers who are not first or third class ticket holders

In [103]:
df[~((df.Pclass == 1) | (df.Pclass == 3))]

Unnamed: 0 Cabin Embarked     Fare  Pclass         Ticket   Age  \
9              9   NaN        C  30.0708       2         237736  14.0   
15            15   NaN        S  16.0000       2         248706  55.0   
17            17   NaN        S  13.0000       2         244373   NaN   
20            20   NaN        S  26.0000       2         239865  35.0   
21            21   D56        S  13.0000       2         248698  34.0   
...          ...   ...      ...      ...     ...            ...   ...   
1278        1278   NaN        S  13.0000       2         244346  57.0   
1284        1284   NaN        S  10.5000       2     C.A. 30769  47.0   
1292        1292   NaN        S  21.0000       2          28664  38.0   
1296        1296   D38        C  13.8625       2  SC/PARIS 2166  20.0   
1297        1297   NaN        S  10.5000       2          28666  23.0   

                                              Name  Parch     Sex  SibSp  \
9              Nasser, Mrs. Nicholas (Adele Achem)      0  female      1   
15                Hewlett, Mrs. (Mary D Kingcome)       0  female      0   
17                    Williams, Mr. Charles Eugene      0    male      0   
20                            Fynney, Mr. Joseph J      0    male      0   
21                           Beesley, Mr. Lawrence      0    male      0   
...                                            ...    ...     ...    ...   
1278                               Ashby, Mr. John      0    male      0   
1284                          Gilbert, Mr. William      0    male      0   
1292                               Gale, Mr. Harry      0    male      1   
1296  Nourney, Mr. Alfred (Baron von Drachstedt")"      0    male      0   
1297                     Ware, Mr. William Jeffery      0    male      1   

      Survived  
9          1.0  
15         1.0  
17         1.0  
20         0.0  
21         1.0  
...        ...  
1278       NaN  
1284       NaN  
1292       NaN  
1296       NaN  
1297       NaN  

[277 rows x 12 columns]

In [104]:
del df['Unnamed: 0']
df.head()

Cabin Embarked     Fare  Pclass            Ticket   Age  \
0   NaN        S   7.2500       3         A/5 21171  22.0   
1   C85        C  71.2833       1          PC 17599  38.0   
2   NaN        S   7.9250       3  STON/O2. 3101282  26.0   
3  C123        S  53.1000       1            113803  35.0   
4   NaN        S   8.0500       3            373450  35.0   

                                                Name  Parch     Sex  SibSp  \
0                            Braund, Mr. Owen Harris      0    male      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      0  female      1   
2                             Heikkinen, Miss. Laina      0  female      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)      0  female      1   
4                           Allen, Mr. William Henry      0    male      0   

   Survived  
0       0.0  
1       1.0  
2       1.0  
3       1.0  
4       0.0

##### IN THIS LESSON
<LI> SELECTING DATA FROM DATAFRAME USING CONDITIONAL DATAFRAME

## Pandas method

#### <ul>
Get descriptive statistics of data

In [105]:
df.describe(include = 'all')

Cabin Embarked         Fare       Pclass    Ticket          Age  \
count           295     1307  1308.000000  1309.000000      1309  1046.000000   
unique          186        3          NaN          NaN       929          NaN   
top     C23 C25 C27        S          NaN          NaN  CA. 2343          NaN   
freq              6      914          NaN          NaN        11          NaN   
mean            NaN      NaN    33.295479     2.294882       NaN    29.881138   
std             NaN      NaN    51.758668     0.837836       NaN    14.413493   
min             NaN      NaN     0.000000     1.000000       NaN     0.170000   
25%             NaN      NaN     7.895800     2.000000       NaN    21.000000   
50%             NaN      NaN    14.454200     3.000000       NaN    28.000000   
75%             NaN      NaN    31.275000     3.000000       NaN    39.000000   
max             NaN      NaN   512.329200     3.000000       NaN    80.000000   

                        Name        Parch   Sex        SibSp    Survived  
count                   1309  1309.000000  1309  1309.000000  891.000000  
unique                  1307          NaN     2          NaN         NaN  
top     Connolly, Miss. Kate          NaN  male          NaN         NaN  
freq                       2          NaN   843          NaN         NaN  
mean                     NaN     0.385027   NaN     0.498854    0.383838  
std                      NaN     0.865560   NaN     1.041658    0.486592  
min                      NaN     0.000000   NaN     0.000000    0.000000  
25%                      NaN     0.000000   NaN     0.000000    0.000000  
50%                      NaN     0.000000   NaN     0.000000    0.000000  
75%                      NaN     0.000000   NaN     1.000000    1.000000  
max                      NaN     9.000000   NaN     8.000000    1.000000

###### <ul>
    we can see in cabin column there are 295 entries with 186 unique values. the most common values are c23 c25 c27 that occurs 6 times
   

In [106]:
type(df.describe(include = 'all'))

pandas.core.frame.DataFrame

#### Splitting, Applying and Combining Data Sources

###### <ul> dataframegroupby object

##### display output of embarked to see what it does

In [107]:
embarked_grouped = df.groupby('Embarked')

In [108]:
embarked_grouped.groups

{'C': Int64Index([   1,    9,   19,   26,   30,   31,   34,   36,   39,   42,
             ...
             1260, 1262, 1266, 1288, 1293, 1295, 1296, 1298, 1305, 1308],
            dtype='int64', length=270),
 'Q': Int64Index([   5,   16,   22,   28,   32,   44,   46,   47,   82,  109,
             ...
             1206, 1249, 1271, 1272, 1279, 1287, 1290, 1299, 1301, 1302],
            dtype='int64', length=123),
 'S': Int64Index([   0,    2,    3,    4,    6,    7,    8,   10,   11,   12,
             ...
             1289, 1291, 1292, 1294, 1297, 1300, 1303, 1304, 1306, 1307],
            dtype='int64', length=914)}

In [109]:
df.iloc[32] # verify to check if the groups are valid

Cabin                            NaN
Embarked                           Q
Fare                            7.75
Pclass                             3
Ticket                        335677
Age                              NaN
Name        Glynn, Miss. Mary Agatha
Parch                              0
Sex                           female
SibSp                              0
Survived                           1
Name: 32, dtype: object

#### lets calculate mean age of each group

In [110]:
for i, j in embarked_grouped:
    print(i, j.Age.mean())

C 32.33216981132075
Q 28.63
S 29.245204603580564


In [111]:
# mean using agg method
embarked_grouped.agg(np.mean)

Fare    Pclass        Age     Parch     SibSp  Survived
Embarked                                                              
C         62.336267  1.851852  32.332170  0.370370  0.400000  0.553571
Q         12.409012  2.894309  28.630000  0.113821  0.341463  0.389610
S         27.418824  2.347921  29.245205  0.426696  0.550328  0.336957

#### Task : Define a simple function that returns the first value in the column and then pass that function through to agg, as an example:


In [112]:
def first_val(x):
    return x.values[8]
embarked_grouped.agg(first_val)

Cabin     Fare  Pclass  Ticket   Age                            Name  \
Embarked                                                                        
C          NaN  11.2417       3    2651  14.0     Nicola-Yarred, Miss. Jamila   
Q          NaN   7.7875       3  330932   NaN  McDermott, Miss. Brigdet Delia   
S         C103  26.5500       1  113783  58.0        Bonnell, Miss. Elizabeth   

          Parch     Sex  SibSp  Survived  
Embarked                                  
C             0  female      1       1.0  
Q             0  female      0       1.0  
S             0  female      0       1.0

In [113]:
df.iloc[11]

Cabin                           C103
Embarked                           S
Fare                           26.55
Pclass                             1
Ticket                        113783
Age                               58
Name        Bonnell, Miss. Elizabeth
Parch                              0
Sex                           female
SibSp                              0
Survived                           1
Name: 11, dtype: object

#### Quantiles
We are interested in knowing the value of dataset below or above which certain fraction of value lie.<br>
df.quantile(0.25, axis = 0)

In [114]:
df.quantile(0.25, axis = 0)

Fare         7.8958
Pclass       2.0000
Age         21.0000
Parch        0.0000
SibSp        0.0000
Survived     0.0000
Name: 0.25, dtype: float64

#### Lambda Functions
One common and useful way of implementing agg is through the use of Lambda functions.

Lambda, or anonymous, functions (also known as inline functions in other languages) are small, single-expression functions that can be declared and used without the need for a formal function definition via the use of the def keyword. Lambda functions are essentially provided for convenience and aren't intended to be used for extensive periods. The main benefit of Lambda functions is that they can be used in places where a function might not be appropriate or convenient, such as inside other expressions or function calls. The standard syntax for a Lambda function is as follows (always starting with the lambda keyword)<br>

In [115]:
embarked_grouped = df.groupby('Embarked')

In [116]:
embarked_grouped.agg(lambda x:x.values[8]) #return first value in column

Cabin     Fare  Pclass  Ticket   Age                            Name  \
Embarked                                                                        
C          NaN  11.2417       3    2651  14.0     Nicola-Yarred, Miss. Jamila   
Q          NaN   7.7875       3  330932   NaN  McDermott, Miss. Brigdet Delia   
S         C103  26.5500       1  113783  58.0        Bonnell, Miss. Elizabeth   

          Parch     Sex  SibSp  Survived  
Embarked                                  
C             0  female      1       1.0  
Q             0  female      0       1.0  
S             0  female      0       1.0

#### we can also pass multiple functions to agg via a list to apply the functions across dataset

In [117]:
embarked_grouped.agg([lambda x: x.values[8], np.mean, np.std])

Fare                           Pclass                      \
         <lambda_0>       mean        std <lambda_0>      mean       std   
Embarked                                                                   
C           11.2417  62.336267  84.185996          3  1.851852  0.936802   
Q            7.7875  12.409012  13.616133          3  2.894309  0.380099   
S           26.5500  27.418824  37.096402          1  2.347921  0.784126   

                Age                            Parch                      \
         <lambda_0>       mean        std <lambda_0>      mean       std   
Embarked                                                                   
C              14.0  32.332170  15.258092          0  0.370370  0.670579   
Q               NaN  28.630000  15.045784          0  0.113821  0.531056   
S              58.0  29.245205  14.047507          0  0.426696  0.943989   

              SibSp                       Survived                      
         <lambda_0>      mean       std <lambda_0>      mean       std  
Embarked                                                                
C                 1  0.400000  0.554803        1.0  0.553571  0.498608  
Q                 0  0.341463  0.885487        1.0  0.389610  0.490860  
S                 0  0.550328  1.161723        1.0  0.336957  0.473037

#### Apply numpy.sum to the Fare column and the Lambda function to the Age column by passing agg a dictionary where the keys are the columns to apply the function to, and the values are the functions themselves to be able to apply different functions to different columns in the DataFrame:


In [118]:
embarked_grouped.agg({
'Fare' : np.sum,
'Age' : lambda x: x.values[0]}) 

Fare   Age
Embarked                  
C         16830.7922  38.0
Q          1526.3085   NaN
S         25033.3862  22.0

#### Finally, execute the groupby method using more than one column. Provide the method with a list of the columns (Sex and Embarked) to groupby, like this:

In [119]:
embarked_group = df.groupby(['Sex', 'Embarked'])

In [120]:
embarked_group.groups # the keys are combination of sex and embarked

{('male',
  'S'): Int64Index([   0,    4,    6,    7,   12,   13,   17,   20,   21,   23,
             ...
             1283, 1284, 1285, 1289, 1292, 1294, 1297, 1304, 1306, 1307],
            dtype='int64', length=623),
 ('female',
  'C'): Int64Index([   1,    9,   19,   31,   39,   43,   52,  111,  114,  128,
             ...
             1238, 1241, 1252, 1255, 1259, 1262, 1266, 1288, 1293, 1305],
            dtype='int64', length=113),
 ('female',
  'S'): Int64Index([   2,    3,    8,   10,   11,   14,   15,   18,   24,   25,
             ...
             1265, 1267, 1273, 1274, 1276, 1282, 1286, 1291, 1300, 1303],
            dtype='int64', length=291),
 ('male',
  'Q'): Int64Index([   5,   16,   46,  116,  126,  143,  171,  188,  196,  214,  245,
              260,  278,  280,  301,  364,  388,  411,  421,  428,  459,  468,
              510,  517,  525,  552,  560,  613,  626,  629,  703,  718,  749,
              768,  776,  778,  787,  790,  825,  828,  890,  891,  893,  907,


#### Data Quality
<li> the quality of data is critical to performance of final model
    <br> Rule of thumb:
    
  <li> if you have clean data
  <li> in sufficient quantity
  <li> with a good correlation between the input data type and the desired output
      <br> then the specifics regarding the type and details of the selected supervised learning model become significantly less important in achieving a good result.<br><br>
In reality, however, this is rarely the case. There are usually some issues regarding the quantity of available data, the quality or signal-to-noise ratio in the data, the correlation between the input and output, or some combination of all three factors. 

##### Data Quality: Problem 1 Managing missing data

In [121]:
complete_data = df.dropna()

##### We can identify which rows are actually missing information and whether the missing information is a problem unique to certain columns or is consistent throughout all columns of the dataset. 

In [122]:
df.agg(lambda x: x.isna().sum())

Cabin       1014
Embarked       2
Fare           1
Pclass         0
Ticket         0
Age          263
Name           0
Parch          0
Sex            0
SibSp          0
Survived     418
dtype: int64

In [123]:
df_valid = df.loc[(~df.Embarked.isna()) & (~df.Fare.isna())]

#### for age column, we can replace all the missing values with the mean, but it is not reasonable to give same value. A better estimate might be considering age as well.

In [138]:
mean_ages = df_valid.groupby(['Pclass','Sex'])['Age'].\
                        transform(lambda x: x.fillna(x.mean()))

In [139]:
df_valid.loc[:,'Age'] = mean_ages

#### Data Quality: Problem 2 Class Imbalance

In [142]:
len(df.loc[df.Survived == 0])

549

#### Techniques: <li> remove samples from the over-represented class until balance has been achieved <li> oversample the under-represented class by randomly copying samples from the under-represented class in the dataset to boost the number of samples<br> Disadvantages <br>While removing data can lead to accuracy issues due to discarding useful information, oversampling the under-represented class can lead to being unable to predict the label of unseen data, also known as overfitting

#### Data Quality: Problem 3 Low Sample Size

In [146]:
SampleRange = range(1,10)
print(len(SampleRange))

9


#### <ul>
A: To replace categorical values, the most frequently occurring value should be used b. 

B: For replacing values where random missing values are used.

C: To replace integer values, the most frequently occurring value should be used .

D: To replace categorical values, the most frequently occurring value should be used .

## ACTIVITY

In [157]:
#load dataset using pandas
df = pd.read_csv('titanic.csv')

In [158]:
df.head()

Unnamed: 0 Cabin Embarked     Fare  Pclass            Ticket   Age  \
0           0   NaN        S   7.2500       3         A/5 21171  22.0   
1           1   C85        C  71.2833       1          PC 17599  38.0   
2           2   NaN        S   7.9250       3  STON/O2. 3101282  26.0   
3           3  C123        S  53.1000       1            113803  35.0   
4           4   NaN        S   8.0500       3            373450  35.0   

                                                Name  Parch     Sex  SibSp  \
0                            Braund, Mr. Owen Harris      0    male      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      0  female      1   
2                             Heikkinen, Miss. Laina      0  female      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)      0  female      1   
4                           Allen, Mr. William Henry      0    male      0   

   Survived  
0       0.0  
1       1.0  
2       1.0  
3       1.0  
4       0.0

In [159]:
# describe the summary data
df.describe()

Unnamed: 0         Fare       Pclass          Age        Parch  \
count  1309.000000  1308.000000  1309.000000  1046.000000  1309.000000   
mean    654.000000    33.295479     2.294882    29.881138     0.385027   
std     378.020061    51.758668     0.837836    14.413493     0.865560   
min       0.000000     0.000000     1.000000     0.170000     0.000000   
25%     327.000000     7.895800     2.000000    21.000000     0.000000   
50%     654.000000    14.454200     3.000000    28.000000     0.000000   
75%     981.000000    31.275000     3.000000    39.000000     0.000000   
max    1308.000000   512.329200     3.000000    80.000000     9.000000   

             SibSp    Survived  
count  1309.000000  891.000000  
mean      0.498854    0.383838  
std       1.041658    0.486592  
min       0.000000    0.000000  
25%       0.000000    0.000000  
50%       0.000000    0.000000  
75%       1.000000    1.000000  
max       8.000000    1.000000

In [160]:
# delete Unnamed: 0 column
df.drop('Unnamed: 0',axis = 1, inplace = True)
df.head()

Cabin Embarked     Fare  Pclass            Ticket   Age  \
0   NaN        S   7.2500       3         A/5 21171  22.0   
1   C85        C  71.2833       1          PC 17599  38.0   
2   NaN        S   7.9250       3  STON/O2. 3101282  26.0   
3  C123        S  53.1000       1            113803  35.0   
4   NaN        S   8.0500       3            373450  35.0   

                                                Name  Parch     Sex  SibSp  \
0                            Braund, Mr. Owen Harris      0    male      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      0  female      1   
2                             Heikkinen, Miss. Laina      0  female      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)      0  female      1   
4                           Allen, Mr. William Henry      0    male      0   

   Survived  
0       0.0  
1       1.0  
2       1.0  
3       1.0  
4       0.0

In [161]:
df.mean()

Fare        33.295479
Pclass       2.294882
Age         29.881138
Parch        0.385027
SibSp        0.498854
Survived     0.383838
dtype: float64

In [162]:
df.median()

Fare        14.4542
Pclass       3.0000
Age         28.0000
Parch        0.0000
SibSp        0.0000
Survived     0.0000
dtype: float64

In [163]:
df.min()

Fare                          0
Pclass                        1
Ticket                   110152
Age                        0.17
Name        Abbing, Mr. Anthony
Parch                         0
Sex                      female
SibSp                         0
Survived                      0
dtype: object

In [164]:
df.max()

Fare                            512.329
Pclass                                3
Ticket                        WE/P 5735
Age                                  80
Name        van Melkebeke, Mr. Philemon
Parch                                 9
Sex                                male
SibSp                                 8
Survived                              1
dtype: object

In [165]:
df.std()

Fare        51.758668
Pclass       0.837836
Age         14.413493
Parch        0.865560
SibSp        1.041658
Survived     0.486592
dtype: float64

In [168]:
df.quantile(0.33)

Fare         8.559325
Pclass       2.000000
Age         23.000000
Parch        0.000000
SibSp        0.000000
Survived     0.000000
Name: 0.33, dtype: float64

In [170]:
df.quantile(0.66)

Fare        26.0
Pclass       3.0
Age         34.0
Parch        0.0
SibSp        0.0
Survived     1.0
Name: 0.66, dtype: float64

In [171]:
df.quantile(0.99)

Fare        262.375
Pclass        3.000
Age          65.000
Parch         4.000
SibSp         5.000
Survived      1.000
Name: 0.99, dtype: float64

In [176]:
df.groupby(df.Pclass)